cleanup: ppm sc api

FedeDP opened the pull request:
**What type of PR is this?**

/kind cleanup

**Any specific area of the project related to this PR?**

/area libscap
/area libsinsp

**Does this PR require a change in the driver versions?**

**What this PR does / why we need it**:

Refactors ppm_sc/tp/events APIs, introducing a new set class by generalizing Falco `falco_event_types` class. 

* properly use ppm_sc and ppme enum (defined as types now) in ppm_sc sinsp API
* added a libsinsp::events:: namespace that holds ppm_sc,tp,events related static functions
* renamed sinsp_{ppm_sc,tp} source files to sinsp_events_{ppm_sc,tp} 
* ported tp enum to a ppm_tp_code type
* added libsinsp::events::set template class

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
refactor(userspace)!: renamed multiple events related API; moreover, added a new libsinsp::events namespace where all the new APIs live. Finally, addeda libsinsp::events::set class to manage the ppm_sc/tp/events types.
```

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/877#" title="Author self-approved">FedeDP</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [FedeDP]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

FedeDP commented:
/cc @jasondellaluce @incertum @Andreagit97

FedeDP commented:
Need to fix libsinsp `interesting_syscalls` and `interesting_tracepoints` tests.

jasondellaluce commented on userspace/libsinsp/CMakeLists.txt:
```diff
@@ -117,8 +117,9 @@ set(SINSP_SOURCES
-	sinsp_ppm_sc.cpp
-	sinsp_tp.cpp)
+	sinsp_events.cpp
+		sinsp_events_ppm_sc.cpp
+		sinsp_events_tp.cpp)
```
Indentation?

FedeDP replied:
Done ;)

jasondellaluce commented on userspace/libsinsp/sinsp.h:
```diff
@@ -850,148 +851,30 @@ class SINSP_PUBLIC sinsp : public capture_stats_source
-		WARNING: without using this method, we cannot guarantee that `libsinsp` state
-		will always be up to date, or even work at all.
-	*/
-	std::unordered_set<uint32_t> enforce_sinsp_state_ppm_sc(std::unordered_set<uint32_t> ppm_sc_of_interest = {});
```
Moving all these is the biggest breaking change of this PR. Please let's make sure the commit names and release notes have a conventional commit prefix containing `!`, and let's be explicit about the action items for fixing the issues of whoever pulls this upstream!

FedeDP replied:
You are right, i will review each commit and add the `!` where needed ;)

FedeDP replied:
Done!

jasondellaluce commented on userspace/libsinsp/sinsp_events.h:
```diff
@@ -0,0 +1,369 @@
+	{
+	}
+
+	inline explicit set(std::unordered_set<ppm_type> vals)
```
Should we add a constructor for `std::set<ppm_type>` so that we have parity?

jasondellaluce commented on userspace/libsinsp/sinsp_events.h:
```diff
@@ -0,0 +1,369 @@
+		return ret;
+	}
+
+	set intersect(const set& other)
```
```suggestion
	set intersect(const set& other) const
```

jasondellaluce commented on userspace/libsinsp/sinsp_events.h:
```diff
@@ -0,0 +1,369 @@
+		return m_types == other.m_types;
+	}
+
+	set diff(const set& other)
```
```suggestion
	set diff(const set& other) const
```

jasondellaluce commented on userspace/libsinsp/sinsp_events.h:
```diff
@@ -0,0 +1,369 @@
+		}
+	}
+
+	void merge(const std::set<ppm_type>& other)
```
Should we one method for `std::unordered_set<ppm_type>` so that we have parity?

jasondellaluce commented on userspace/libsinsp/sinsp_events.h:
```diff
@@ -0,0 +1,369 @@
+	}
+
+	set intersect(const set& other)
+	{
```
I know we have `merge` that does a similar thing, but at this point would it make sense to add `set union(const set& other) const`? That way this class would be all-around versatile.

FedeDP replied:
Yep fully agree!

jasondellaluce commented on userspace/libsinsp/sinsp_events.h:
```diff
@@ -0,0 +1,369 @@
+	WARNING: without using this method, we cannot guarantee that `libsinsp` state
+	will always be up to date, or even work at all.
+*/
+set<ppm_sc_code> enforce_sinsp_state_ppm_sc(set<ppm_sc_code> ppm_sc_of_interest = {});
```
Since we are refactoring this bottom-up, I was thinking that maybe we could make these function names shorter, add the few ones missing, and be more consistent. Here's my proposal, but feel free to ignore it as this is more a matter of taste (the bold ones have not been implemented so far):
- all_sc_set
- all_tp_set
- **all_event_set**
- sinsp_state_sc_set
- sinsp_state_tp_set
- sinsp_state_event_set
- simple_sc_set
- io_sc_set
- io_sc_set
- file_sc_set
- net_sc_set
- proc_sc_set
- sys_sc_set
- sc_set_to_names
- tp_set_to_names
- event_set_to_names
- names_to_sc_set
- **names_to_tp_set**
- **names_to_event_set**
- sc_set_to_event_set
- **event_set_to_sc_set**

incertum replied:
Agree that `ppm` and `ppme` can be reserved as a concept under the hood, and in the API we use `sc` and `event` only. However, the `enforce` prefix would make it clearer that you can pass a set and add to it, so I would keep that.

FedeDP replied:
> However, the enforce prefix would make it clearer that you can pass a set and add to it, so I would keep that.

Given that new `libsinsp::events::set` class provides `merge` method, we could just "enforce" the set for `enforce_simple_ppm_sc_set`, leaving all other sets as just ... sets, without enforcing anything.
Libs client could do something like:
```
io_set = libsinsp::events::io_sc_set();
net_set = libsinsp::events::net_sc_set();

final_set = io_set.merge(net_set);
// use final set
```

FedeDP replied:
@jasondellaluce love the renamings! Will leave them as latest commit, once all build issues are fixes (macos... :/ )

jasondellaluce replied:
Love the approach of using `merge`! With that, most of these sets may also be made static so that we only need to built them at start time, and reuse them for all the runtime calls.

FedeDP replied:
Note: i'd avoid adding any new API for now; just rename existing ones; we can later add them as we need.
(eg: `libsinsp::events::all_event_set` would be new)

FedeDP replied:
Latest commit updated all function names!

jasondellaluce reviewed (commented):
Love all this! It's now much more clear!

I left you few initial comments, feel free to ignore some of those if you think they're out of context

FedeDP commented:
@jasondellaluce NOTE: i am still open to update APIs namings, since now we have a proper namespace and proper types.

incertum commented on userspace/libsinsp/sinsp_events_ppm_sc.cpp:
```diff
@@ -0,0 +1,309 @@
+	std::vector<uint32_t> minimum_syscalls(PPM_SC_MAX, 0);
+
+	/* Should never happen but just to be sure. */
+	if(scap_get_modifies_state_ppm_sc(minimum_syscalls.data()) != SCAP_SUCCESS)
```
adding to @jasondellaluce point, given we refactor from the ground up, would love to see `scap_get_modifies_state_ppm_sc` be defined in this API and if you agree @FedeDP we might as well do the refactor of only enforcing the EF_MODIFIES_STATE and drop the UF_NEVER_DROP enforcement, but this can maybe be done in a follow up PR to not overload this refactor. Also it's more sinsp modifies state ... so we could rename one more time as well.

FedeDP replied:
> we might as well do the refactor of only enforcing the EF_MODIFIES_STATE and drop the UF_NEVER_DROP enforcement

Yep, i'd leave this to another PR since it is a somewhat disruptive change!

>  would love to see scap_get_modifies_state_ppm_sc be defined in this API

You mean a `sinsp_modifies_state_set`? This is what `enforce_sinsp_state_ppm_sc` already does, right?

incertum replied:
Re your q: `sinsp_state_sc_set()` calls `scap_get_modifies_state_ppm_sc` and not sure why `scap_get_modifies_state_ppm_sc` lives in scap. It appears that we could move all the code directly to `sinsp_state_sc_set()` which may not only simplify everything, but also increase the clarity wrt sinsp <-> scap responsibility separation, something Jason and others brought up too in past threads.

We can chat more about it in the follow up PR :)

FedeDP replied:
This would make a lot of sense; BUT we first would need to drop `UF_NEVER_DROP` check, since syscall table is not exposed to sinsp ;) 
We can do it later once we decide dropping `UF_NEVER_DROP` set is something we actually want!

incertum replied:
Yup you probably already know my opinion on `UF_NEVER_DROP` :upside_down_face: would urge to get rid of it for the syscalls selection to have proper separation.

FedeDP replied:
Well, `UF_NEVER_DROP` is not taken into account unless you enable dropping mode ;) but Falco was just cleaned up about that: https://github.com/falcosecurity/falco/pull/2415

incertum reviewed (commented):
Agree with making all set operations complete to cover all future use cases.

FedeDP commented on driver/bpf/plumbing_helpers.h:
```diff
@@ -301,7 +301,7 @@ static __always_inline bool is_syscall_interesting(int id)
 }
 
-static __always_inline const struct ppm_event_info *get_event_info(enum ppm_event_type event_type)
+static __always_inline const struct ppm_event_info *get_event_info(ppm_event_code event_type)
```
It would make sense! But we already have:
* ppm_event_category
* ppm_event_entry
* ppm_event_flags
* ppm_event_info

:P

incertum replied:
I know, we can live with it :)

FedeDP commented on userspace/libsinsp/sinsp_events.h:
```diff
@@ -0,0 +1,397 @@
+
+	inline size_t size() const
+	{
+		return len;
```
Done ;)

FedeDP commented on userspace/libscap/userspace_flag_helpers.h:
```diff
@@ -10,6 +10,7 @@
 #include <sys/ptrace.h>
 #include <sys/resource.h>
 #include <sys/file.h>
+#include <sched.h>
```
This fixed issues building unit tests on my archlinux laptop.

incertum commented:
It's amazing how much more clarity this refactor and renaming achieves, well done 👏 ! Also nice addition `get_sc_set_from_cat`!

@FedeDP may I kindly ask you to consider adding two more mapping functions?
- `names_to_event_set` no current use case, but let's try to be future proof if possible
- `event_set_to_sc_set` adding this as well may give more options to convert the evt types from the filters to state sets

---

Would say let's move forward and get this out of WIP. @Andreagit97 @jasondellaluce @FedeDP what additional testing would you suggest before merging? From my perspective it looks already pretty ready, but let's be extra careful with this critical refactor.

FedeDP commented:
> may I kindly ask you to consider adding two more mapping functions?

Implemented, together with `all_event_set` ;) 

Moreover, updated scap.h related APIs to take an `uint8_t` array (instead of `uint32_t`), and optimized their calls by adding a new `libsinsp::events::set.data()` method that returns underlying `uint8_t` vector data.

FedeDP commented:
Latest commit added unit tests for `libsinsp::events::set` class.

Andreagit97 commented on userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp:
```diff
@@ -271,59 +271,40 @@ std::set<uint32_t> ordered_sinsp_state_ppm_sc_set{
+	libsinsp::events::set<ppm_sc_code> additional_syscalls;
+	auto ppm_sc_matching_set = expected_sinsp_state_ppm_sc_set;
 
 #ifdef __NR_kill
```
At the end of our work we should remove all these `ifdef`, we should be able to work on PPM_SC codes without considering the running architecture :tada:

Andreagit97 reviewed (approved):
/approve
Amazing work @FedeDP !!

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/877#pullrequestreview-1297629740" title="Approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/877#" title="Author self-approved">FedeDP</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,FedeDP]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

FedeDP commented on userspace/libsinsp/sinsp_events.h:
```diff
@@ -0,0 +1,430 @@
+  It has same effect of old `simple_consumer` mode.
+  Does enforce minimum sinsp state set.
+*/
+set<ppm_sc_code> enforce_simple_sc_set(set<ppm_sc_code> ppm_sc_set = {});
```
Nope, this is the `enforce` one (i decided to leave a `enforce_simple_X_set` under each {tp,sc} categories; i think that a consumer should still be able to use `enforce_simple_sc_set(mySet)` and forget about anything else.

Andreagit97 replied:
right now I would keep them, let's say these are our suggested sets to use for libs consumers. Falco won't probably use these APIs anymore.

incertum replied:
Either way fine with me since we now generally adopt an approach of leveraging generic set operations. As we are now breaking everything anyways my preference would be to go all in and just return the simple_set etc in the sinsp API -> that way one consistent behavior and approach across the board.

FedeDP commented on userspace/libsinsp/sinsp_events.cpp:
```diff
@@ -0,0 +1,157 @@
+	libsinsp::events::set<ppm_event_code> ppm_event_set;
+	for (int ppm_ev = 0; ppm_ev < PPM_EVENT_MAX; ++ppm_ev)
+	{
+		std::string ppm_ev_name = g_infotables.m_event_info[ppm_ev].name;
```
Agree!

FedeDP replied:
I implemented a different logic to avoid adding any new API to sinsp (or continuously calling `scap_get_events_from_ppm_sc` in a loop).

Latest commit implements it + adds a test and + moves all libsinsp::events:: API under libsinsp/events/ folder.

FedeDP commented on userspace/libsinsp/events/sinsp_events.cpp:
```diff
@@ -115,21 +137,11 @@ libsinsp::events::set<ppm_event_code> libsinsp::events::all_event_set()
-	});
-
-	if(scap_get_ppm_sc_from_events(events_array.data(), ppm_sc_set.data()) != SCAP_SUCCESS)
+	if(scap_get_ppm_sc_from_events(events_of_interest.const_data(), ppm_sc_set.data()) != SCAP_SUCCESS)
```
Directly pass `events_of_interest` underlying vector as const.

FedeDP commented on userspace/libsinsp/events/sinsp_events_ppm_sc.cpp:
```diff
@@ -240,17 +240,8 @@ libsinsp::events::set<ppm_sc_code> libsinsp::events::names_to_sc_set(const std::
-	});
-
-	if(scap_get_events_from_ppm_sc(ppm_sc_array.data(), events_set.data()) != SCAP_SUCCESS)
+	if(scap_get_events_from_ppm_sc(ppm_sc_set.const_data(), events_set.data()) != SCAP_SUCCESS)
```
Directly pass `ppm_sc_set` underlying vector as const.

Andreagit97 reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/877#pullrequestreview-1297996981" title="Approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/877#" title="Author self-approved">FedeDP</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,FedeDP]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

incertum commented on userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp:
```diff
@@ -407,15 +374,17 @@ TEST(interesting_syscalls, get_syscalls_names)
 	orderd_syscall_names_matching_set.insert("read");
 
-	auto orderd_syscall_names_final_set = test_utils::unordered_set_to_ordered(inspector->get_syscalls_names(ppm_sc_set));
+	auto syscall_names_final_set = libsinsp::events::sc_set_to_names(ppm_sc_set);
```
Add one more test re `names_to_sc_set` and include one non syscall event, e.g. `container` in names as well to test that we only retrieve the correct `sc_set`.

FedeDP replied:
I think we can improve all of these in a subsequent PR ;) we will be working around the new API anyway!

incertum reviewed (commented):
Focused on scrutinizing the test conditions and left a few comments.

incertum commented:
/hold

incertum reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/877#pullrequestreview-1297996981" title="Approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/877#" title="Author self-approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/877#pullrequestreview-1298506625" title="Approved">incertum</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,FedeDP,incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

FedeDP commented:
/unhold

FedeDP commented:
/milestone 0.11.0
